Add file reader and writer for protobuf messages

felippe-mendonca opened the pull request:
Some stuff can be done to improve, e.g.:
- return an error code (is::wire::Status) when file can't be opened in both _ProtobufReader_ and _ProtobufWriter_
- maybe, on _ProtobufReader::next_, returns is::wire::Status and pass a reference to desired message, as well as in _ProtobufReader::insert_ method
- if it's possible, return an error code on _ProtobufReader::next_ when isn't possible to desarealize the message for the given schema 

I'm open for discuss the modelling of these features!

felippe-mendonca committed:
+ classes to read and write protobuf messages to file

rodolfo-picoreti commented on is/utils/io.hpp:
```diff
@@ -1,10 +1,64 @@
 #include "status.hpp"
+#include "boost/optional.hpp"
+
+namespace google {
```
Protobuf is already a public dependency, dont worry about forward declaring this

rodolfo-picoreti commented on is/utils/io.hpp:
```diff
@@ -1,10 +1,64 @@
+
+struct ProtobufWriter {
+  std::ofstream file;
+  google::protobuf::io::ZeroCopyOutputStream *raw_output;
```
Use OstreamOutputStream instead of the interface to avoid unnecessary dynamic allocation?

felippe-mendonca replied:
I tried to use smart pointers but it seems that the buffer weren't flush (read the generated file with a python script). I'm going to post this issue on this [stackoverflow](https://stackoverflow.com/questions/2340730/are-there-c-equivalents-for-the-protocol-buffers-delimited-i-o-functions-in-ja) question that I've based for this last commit.

rodolfo-picoreti commented on is/utils/io.hpp:
```diff
@@ -1,10 +1,64 @@
+
+struct ProtobufReader {
+  std::ifstream file;
+  google::protobuf::io::ZeroCopyInputStream *raw_input;
```
Use IstreamInputStream instead of the interface to avoid unnecessary dynamic allocation?

rodolfo-picoreti commented on is/utils/io.cpp:
```diff
@@ -68,4 +70,36 @@ wire::Status save(std::string const& filename, google::protobuf::Message const&
+  }
+}
+
+ProtobufReader::ProtobufReader(std::string const &filename)
```
Example constructor without dynamic allocation:
```
ProtobufReader::ProtobufReader(std::string const &filename)
    : file(filename, std::ios::in | std::ios::binary),
      raw_input(&file),
      coded_input(&raw_input) {
}

```

rodolfo-picoreti commented on is/utils/io.hpp:
```diff
@@ -1,10 +1,64 @@
+  ~ProtobufWriter();
+  void close();
+
+  template <typename T> bool insert(T const &message) {
```
Maybe change the signature to return Status? And return a FAILED_PRECONDITION if we failed to open the file before, otherwise throw at the class constructor...

rodolfo-picoreti commented on is/utils/io.hpp:
```diff
@@ -1,10 +1,64 @@
+  ~ProtobufReader();
+  void close();
+
+  template <typename T> boost::optional<T> next() {
```
Same comments as in ProtobufWrite::insert. If you decide to return Status here change function signature to receive google::protobuf::Message* (and to the same for ProtobufWrite::insert)

felippe-mendonca committed:
improvements on Writer/Reader classes
- ProtobufWriter::insert and ProtobufReader::next now return an wire::Status
- remove foward declarations from io.hpp and put the includes on it
- tests were remade to fit with new implementation
